### PR TITLE
act: avoid ReceiveTimeout handling in hot path if no receive timeout is set

### PR DESCRIPTION
**Pull request by jrudolph**
@patriknw can you run the ActorBenchmark with and without this change? For me it makes more than 10% difference. It seems `isInstanceOf` checks are quite costly in the hot path.

### CODE REVIEW

**Comment by patriknw** [comment]
ok, I'll run it now

**Comment by patriknw** [comment]
s/cores/threads/

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 1]
```diff
@@ -24,6 +26,10 @@ private[akka] trait ReceiveTimeout { this: ActorCell ⇒
 
   final def setReceiveTimeout(timeout: Duration): Unit = receiveTimeoutData = receiveTimeoutData.copy(_1 = timeout)
 
+  protected def checkReceiveTimeoutIfNeeded(message: Any): Unit =
+    if (hasTimeoutData)
+      checkReceiveTimeout(!message.isInstanceOf[NotInfluenceReceiveTimeout])
+
   final def checkReceiveTimeout(reschedule: Boolean = true): Unit = {
     val (recvtimeout, task) = receiveTimeoutData
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: I'm surprised that instanceof would have such big impact, could it be and the subsequent checks that is the culprit?

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 2]
```diff
@@ -24,6 +26,10 @@ private[akka] trait ReceiveTimeout { this: ActorCell ⇒
...
     recvtimeout match {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: We have had a few issues with receive timeout so we must be sure that it's doing the same, see below comment... as far as I can see it's correct.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala (hunk 1 above)
Comment (jrudolph) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: It's probably the whole constructing and deconstructing of the tuple. It's a bit hard to say exactly from looking at jmh's `-prof perfasm` output...

The instanceOf calls are definitely not for free, though. The `AutoReceivedMessage` filtering even if never used during the benchmark is quite expensive as well (but I don't see a way around it).

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala (hunk 2 above)
Comment (jrudolph) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: Yep, needs proper review. I tried to make sure not to break something.

**Comment by jrudolph** [comment]
Failure was a 2.11 mima problem. Fixed.

**Review by johanandren — APPROVED** [review]
Been staring at this for a while now, and I also think it is correct. But maybe a third pair of eyes wouldn't hurt in this case.

LGTM

**Review by raboof — COMMENTED** [review]
The logic looks correct to me, but now while this PR saves an `instanceOf` when `!hasTimeoutData` before the message handling, it also adds one when `hasTimeoutData` after the message handling.

Perhaps it is still worth it to collapse those 2 into 1 for the case `hasTimeoutData` is true both before and after message handling (at the expense of adding a local variable)

**Comment by johanandren** [comment]
Doesn't it make sense that the overhead of the receive timeout is only there if you have a timeout?

**Comment by raboof** [comment]
Of course! I just wondered if we could do even better.

**Comment by jrudolph** [comment]
> The logic looks correct to me, but now while this PR saves an `instanceOf` when `!hasTimeoutData` before the message handling, it also adds one when `hasTimeoutData` after the message handling.
> 
> Perhaps it is still worth it to collapse those 2 into 1 for the case `hasTimeoutData` is true both before and after message handling (at the expense of adding a local variable)

The reasoning is to make the fast path fast while not over-optimize the slow path. Scheduling/cancelling a receive timeout is not cheap anyway so removing an `instanceof` seems not worthwhile in that case.

Looking a bit more at the code in `ReceiveTimeout` I'm not even completely sure it always does the right thing if `!message.isInstanceOf[NotInfluenceReceiveTimeout]` because it doesn't seem to take into account what happens if the timeout is *changed* when a message should `NotInfluenceReceiveTimeout` (created https://github.com/akka/akka/issues/26880 to clarify).

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala (hunk 1 above)
Comment (helena) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: I can run as a pattern match 840 times in the time it takes to execute one `isInstanceOf`

**Pull request merged**